File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
n, m = map(int,input().split())

visit = [True] * n
num = []
res = []

def dfs(visit, num, depth, at):
    if depth == m:
        for i in num:
            print(i, end= ' ')
        print()
        return

    for i in range(at, n):
        if visit[i]:
            visit[i] = False
            num.append(i+1)
            dfs(visit, num, depth+1, i+1)

            visit[i] = True
            num.pop()

dfs(visit, num, 0, 0)

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    static boolean[] visit;
    static int[] list;
    static List<Integer> num;
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String str = bf.readLine();
        num = new ArrayList<>();

        for(String st : str.split(" ")){
            num.add(Integer.parseInt(st));
        }
        visit = new boolean[num.get(0)+1];
        list = new int[num.get(1)];

        permutation(0, 1);
    }

    static void permutation(int count, int start){
        if(count == num.get(1)){
            for(int i : list){
                System.out.print(i + " ");
            }
            System.out.println();
            return;
        }
        for(int i = start ; i <= num.get(0) ; i++){
            if(!visit[i]){
                visit[i] = true;
                list[count] = i;
                permutation(count+1 , i);
                visit[i] = false;
            }
        }
    }

}
